### PR TITLE
feat(Degeneration): Added the inspection of degenerated polygons in S…

### PR DESCRIPTION
**Pull request by MelchiorSchuh**
…urfaceMeshes and polyhedra in SolidMeshes.

### CODE REVIEW

File: include/geode/inspector/criterion/orientation/surface_orientation.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    FORWARD_DECLARATION_DIMENSION_CLASS( SurfaceMesh );
+} // namespace geode
+
+namespace geode
+{
+    /*!
+     * Class for inspecting the orientation of the polygons of a SurfaceMesh
+     */
+    template < index_t dimension >
+    class opengeode_inspector_inspector_api SurfaceMeshAdjacency
```
Comment (panquez) [inline include/geode/inspector/criterion/orientation/surface_orientation.h:41]: does it check orientations or adjacencies?

File: src/geode/inspector/criterion/degeneration/solid_degeneration.cpp [hunk 2]
```diff
@@ -41,6 +44,83 @@ namespace geode
...
+                const auto& p2 = mesh.point(
+                    mesh.polyhedron_vertex( { polyhedron_id, i } ) );
+                const auto& p3 =
+                    mesh.point( mesh.polyhedron_vertex( { polyhedron_id,
+                        static_cast< local_index_t >( i + 1 ) } ) );
+                const auto& p4 =
+                    mesh.point( mesh.polyhedron_vertex( { polyhedron_id,
+                        static_cast< local_index_t >( i + 2 ) } ) );
+                const Tetrahedron polyhedron_tetra{ p1, p2, p3, p4 };
+                if( tetrahedron_volume( polyhedron_tetra ) > global_epsilon )
```
Comment (panquez) [inline src/geode/inspector/criterion/degeneration/solid_degeneration.cpp:112]: is it true? Imagine a perfect cube. You can take 4 vertices on the same plane (same face) so the volume is equal to 0, but the polyhedron is not degenerated.

File: src/geode/inspector/criterion/degeneration/solid_degeneration.cpp [hunk 3]
```diff
@@ -41,6 +44,83 @@ namespace geode
...
+                {
+                    return false;
```
Comment (panquez) [inline src/geode/inspector/criterion/degeneration/solid_degeneration.cpp:114]: Maybe I will check the sum of signed volumes of decomposing tetrahedra (i.e. ~ the volume of the polyhedron).

File: src/geode/inspector/criterion/degeneration/surface_degeneration.cpp [hunk 4]
```diff
@@ -41,6 +44,75 @@ namespace geode
...
+                mesh.point( mesh.polygon_vertex( { polygon_id, 0 } ) );
+            for( const auto i :
+                LRange{ 1, mesh.nb_polygon_vertices( polygon_id ) - 1 } )
+            {
+                const auto& p2 =
+                    mesh.point( mesh.polygon_vertex( { polygon_id, i } ) );
+                const auto& p3 = mesh.point( mesh.polygon_vertex(
+                    { polygon_id, static_cast< local_index_t >( i + 1 ) } ) );
+                const Triangle< dimension > polygon_triangle{ p1, p2, p3 };
+                if( triangle_area( polygon_triangle ) > global_epsilon )
```
Comment (panquez) [inline src/geode/inspector/criterion/degeneration/surface_degeneration.cpp:105]: Same question here.

**Review by panquez — APPROVED** [review]

File: src/geode/inspector/criterion/degeneration/surface_degeneration.cpp (hunk 4 above)
Comment (BotellaA) [inline src/geode/inspector/criterion/degeneration/surface_degeneration.cpp:105]: Not sure about that in the case of a concave polygon...

**Review by BotellaA — APPROVED** [review]

**Pull request merged**